### ISSUE TITLE
Errors in templates are swallowed by template lookup plugin

### ISSUE THREAD

**Issue by scottanderson42** [ansible/ansible#7370]
##### Issue Type:

Bug Report
##### Ansible Version:

ansible 1.7 (devel a6a1e240c6) last updated 2014/05/12 10:59:56 (GMT -400)

This happens in 1.6 as well.
##### Environment:

OSX 10.9.2 
##### Summary:

Errors in templates are not propagated when using the template lookup plugin, which returns an empty string instead. Errors in these templates can be very difficult to debug as a result.
##### Steps To Reproduce:

ansible-playbook -vvv -i inventory/local pb_test.yml

https://gist.github.com/scottanderson42/c1d172aa56916a258ff1
##### Expected Results:

The first task should fail as the variable does_not_exist is not set.
##### Actual Results:

The lookup task returns an empty string when the variable is not set instead of failing.

(sp_devops)GoatRodeo:devops anderson (development)$ ansible-playbook -vvv -i inventory/local pb_test.yml

PLAY [127.0.0.1] **************************************************************

TASK: [debug msg=] ************************************************************
ok: [127.0.0.1] => {
    "msg": ""
}

PLAY [127.0.0.1] **************************************************************

TASK: [debug msg=This does not exist: 42
] ************************************
ok: [127.0.0.1] => {
    "msg": "This"
}

PLAY RECAP ********************************************************************
127.0.0.1                  : ok=2    changed=0    unreachable=0    failed=0

**Comment by 1player** [comment]
I think the error is not only on the template lookup, I'm hitting the same problem with the file lookup plugin.

My playbook has:

```
    - name: deploy key
      authorized_key: user=user key="{{ lookup('file', '/not/existing') }}"
```

When I run it I get an ok from Ansible, even though the file lookup plugin raises an AnsibleError exception if the path does not exist.

I guess the jinja interpolation "{{ function }}" swallows the exception if function throws.

**Comment by piffey** [comment]
Just confirming that this seems to be a lookup plugin issue. Just noticed that a homemade lookup plugin's errors were being swallowed. I'm guessing it's in utils/template.py where ran is being set to None under the lookup method if there's an exception, then the plugin is just returning None. I'll hopefully have a moment to fix this and test it, but just a heads up if anyone else gets to bug squashing first.

**Comment by AlexRiedler** [comment]
:+1: also running into this error, very painful when your setting a configuration and it puts a blank file in for it due to an undefined variable for instance.

**Comment by jgumbley** [comment]
also have this issue :+1: very subtle one to debug

**Comment by romanrev** [comment]
Indeed a very annoying one, already hit it more than once and it always drains a lot of time and effort to debug. :+1:

**Comment by Ernest0x** [comment]
:+1:  Came across this bug while writing a custom lookup plugin on 1.7.1

**Comment by fschulze** [comment]
The solution seems to let AnsibleErrors pass on, for example in the template lookup function like this::

```
def lookup(name, *args, **kwargs):
    from ansible import utils
    instance = utils.plugins.lookup_loader.get(name.lower(), basedir=kwargs.get('basedir',None))
    vars = kwargs.get('vars', None)

    if instance is not None:
        # safely catch run failures per #5059
        try:
            ran = instance.run(*args, inject=vars, **kwargs)
        except errors.AnsibleError:
            raise
        except Exception, e:
            ran = None
        if ran:
            ran = ",".join(ran)
        return ran
    else:
        raise errors.AnsibleError("lookup plugin (%s) not found" % name)
```

**Comment by rngadam** [comment]
I spent a lot of time twice now tracking down the problem to lookup returning an empty string instead of the file content... +1

**Comment by bcoca** [comment]
This is fixed in current dev (probably by changes in 1.8.2) tested by running playbook in gist report:

ERROR: unable to read /home/bcoca/tmp/templates/foo.txt

**Comment by romanrev** [comment]
Thank you, Brian @bcoca! I am just guessing the comments were swapped - the one above should go to #8619  :-)
